pkg/bitcoin/electrum/electrum_integration_test.go

```diff
@@ -1,119 +1,156 @@
+			RequestRetryTimeout: timeout * 6,
 		},
+		network: bitcoin.Testnet,
 	},
 	"electrumx wss": {
-		clientConfig: Config{
+		clientConfig: electrum.Config{
 			URL:                 "wss://electrumx-server.test.tbtc.network:8443",
 			RequestTimeout:      timeout,
 			RequestRetryTimeout: timeout * 2,
 		},
-		errorMessages: expectedErrorMessages{
-			missingTransaction:        "errNo: 2, errMsg: daemon error: DaemonError({'code': -5, 'message': 'No such mempool or blockchain transaction. Use gettransaction for wallet transactions.'})",
-			missingBlockHeader:        "errNo: 1, errMsg: height 4,294,967,295 out of range",
-			missingTransactionInBlock: "errNo: 1, errMsg: tx aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa not in block at height 123,456",
-		},
+		network: bitcoin.Testnet,
 	},
 	"fulcrum tcp": {
-		clientConfig: Config{
-			URL:                 "tcp://testnet.aranguren.org:51001",
+		clientConfig: electrum.Config{
+			URL:                 "tcp://blackie.c3-soft.com:57005",
 			RequestTimeout:      timeout,
 			RequestRetryTimeout: timeout * 2,
 		},
-		errorMessages: expectedErrorMessages{
-			missingTransaction:        "errNo: 2, errMsg: daemon error: DaemonError({'code': -5, 'message': 'No such mempool or blockchain transaction. Use gettransaction for wallet transactions.'})",
-			missingBlockHeader:        "errNo: 1, errMsg: Invalid height",
-			missingTransactionInBlock: "errNo: 1, errMsg: No transaction matching the requested hash found at height 123456",
-		},
+		network: bitcoin.Testnet,
 	},
 }
 
+var invalidTxID bitcoin.Hash
+
+//go:linkname readEmbeddedServers github.com/keep-network/keep-core/config.readElectrumUrls
+func readEmbeddedServers(network bitcoin.Network) ([]string, error)
+
+func init() {
+	var err error
+
+	readServers := func(network bitcoin.Network) error {
+		servers, err := readEmbeddedServers(network)
+		if err != nil {
+			return err
+		}
+
+		for _, server := range servers {
+			serverName := fmt.Sprintf("embedded/%s/%s", network.String(), server)
+			testConfigs[serverName] = testConfig{
+				clientConfig: electrum.Config{
+					URL:                 server,
+					RequestTimeout:      timeout * 2,
+					RequestRetryTimeout: timeout * 4,
+				},
+				network: network,
+			}
+		}
+		return nil
+	}
+
+	if err := readServers(bitcoin.Testnet); err != nil {
+		panic(err)
+	}
+
+	if err := readServers(bitcoin.Mainnet); err != nil {
+		panic(err)
+	}
+
+	// Remove duplicates
+	urls := make(map[string]string)
+	for key, server := range testConfigs {
+		firstName, ok := urls[server.clientConfig.URL]
+		if ok {
+			delete(urls, key)
+			fmt.Printf(
+				"removed server [%s] as a server with the same URL [%s] is already registered under [%s] name\n",
+				key,
+				server.clientConfig.URL,
+				firstName,
+			)
+			continue
+		}
+		urls[server.clientConfig.URL] = key
+	}
```

Comment:
I am confused about this piece of code. Servers in `urls` do not contain duplicate URLs but `urls` is not used anywhere else. 

I would be comfortable removing this loop altogether given duplicating entries between `testConfigs` in this file and embedded in the config files is a low-risk and should be caught during a review. This is not a normal situation to duplicate them.

Reply:
It should remove entry from `testConfigs`. It's fixed here: edc100a7c34e95ea01d5f07dff5112b405d6c244.

The duplicates may happen, we want to maintain the two lists (servers defined here in `electrum_integration_test.go` and in embedded servers) independently. With this solution, we don't need to remember about catching duplicates.